### PR TITLE
docs: blog - deploying wave apps on heroku

### PR DESCRIPTION
**Pull request by srini-x**
A blog post explaining how to deploy Wave apps on Heroku.

 In response to this Stack Overflow question
https://stackoverflow.com/questions/65836484/deploy-h2o-wave-application-to-heroku

### CODE REVIEW

**Review by mturoci — CHANGES_REQUESTED** [review]
Thanks @srini-x, the blog post looks great.

I would suggest simplifying the docker part though and cover only the simplest happy path (0..20+). The reason is, although correct, your proposed deployment is too complex. The primary consumers of this blog post will be people that have no idea how docker or linux in general works (if they did, they wouldn't need this in the first place) so we should give them something they are able to understand and quickly add into their apps. The simplest dockerfile I can think of is:

```dockerfile
# Get base image with python installed.
FROM python:3.8
# Copy over everything.
COPY . .
# Install needed dependencies.
RUN pip install -r requirements.txt
# Run the app.
CMD ["wave", "run", "--no-reload", "app.py"]
```

So the guide should be something like:

* Create such dockerfile
* Adjust your desired python version if necessary (3.8 should work fine)
* Make sure you have requirements.txt present
* Adjust your python module name in `CMD` (instead of app.py)
* Build it using docker build ...
* Verify it works using docker run ...

**Comment by srini-x** [comment]
@mturoci I agree with you on the idea that the target audience is not very familiar with Docker or Linux. That is the reason I had to make this Dockerfile.

with this setup, users can use any version of Python or Wave any Python module name and they don't have to edit any files

steps are actually very simple

1. copy `Dockerfile`, `docker_entrypoint.sh`. we can even exclude `.dockerignore`
2. run 

```shell
$ docker build \
 --build-arg PYTHON_VERSION=3.8 \
 --build-arg WAVE_VERSION=0.20.0 \
 --build-arg PYTHON_MODULE="ai_app/app.py" \
 -t wave-heroku:0.1.0 .
```
3. Done

Also, the Dockerfile need to be a bit more complicated than what you suggested because `$PORT` is randomly selected by Heroku during the time of deployment.

so, we need to add `H2O_WAVE_ADDRESS="http://127.0.0.1:${PORT}"` and `-listen ":${PORT}"` flag to `waved`
I think we can use environment variable `H2O_WAVE_LISTEN` instead of the `-listen` flag.

Since `$PORT` is defined during runtime, we need the `docker-entrypoint.sh`.

**Comment by mturoci** [comment]
> with this setup, users can use any version of Python or Wave any Python module name and they don't have to edit any files

Ye, I agree. My point is more about the understanding of the copy-pasted code. If it works for them, fine, but if it will require some additional minor tweaks, there is no chance they can make it work as it's too complex. People who are able to understand it don't need this blog anyway as they can write it up themselves.

I consider this blog post a "toy-deployment" for "toy-apps", which means for people who have minimal programming skills but still would like to showcase their app somehow. Real prod deployments probably won't be made by a data scientist.

> If you feel strongly that the Dockerfile need to be simplified, I can make a new simplified version that works with $PORT

Just wondering what would be the best option for our target audience. Let's also invite @lo5 and @mtanco to the discussion. You can leave it as it is until we make a final decision.

### Final note

It would be good to also add a link to HAC at the end. Something in the lines of - if you find this too complex/not so scalable, try out our HAC for enterprises which makes this whole process way easier etc.

**Comment by srini-x** [comment]
@mturoci cool I agree with you.

**Comment by woobe** [comment]
IMHO, readers may not want to read the detailed solution straightaway. For this blog post, I would recommend the following:

First, we can start with the problem statement (in this case, what are the challenges of deploying Wave apps).

After that, we can talk about a few possible solutions:
- Solution 1: Heroku, pros and cons, final outcomes (with a link to the full step-by-step instructions)
- Solution 2: Another solution, pros and cons, etc
- ...
- Solution x: Finally, the beauty of HAIC, final outcomes (with some cool screenshots showing how to import apps)

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+$ cd wave-heroku
+```
+
+#### Setup Virtual Environment
+
+```shell
+$ python -V
+Python 3.8.11
+
+$ python -m venv .venv
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:60]:
```suggestion
$ python -m venv venv
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+$ . .venv/bin/activate
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:61]:
```suggestion
$ ./venv/bin/activate
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+author_title: Sr. AI Software Engineer @ H2O.ai
+author_url: https://github.com/srini-x
+author_image_url: https://media-exp1.licdn.com/dms/image/C5603AQFDvAMk4rSJHQ/profile-displayphoto-shrink_400_400/0/1516346833817?e=1648684800&v=beta&t=45kV7SbqzTUDznPYGLAftGzrW5GC-hUfD6OFM3Tbe20
+image: /img/wave-type-yellow-1100x400.png
+
+tags: [community-content]
+---
+
+Wave makes it very easy to create data science applications. Heroku is a cloud platform where we can quickly deploy our Wave app and share it with the rest of the world.
+The following is a walk through on how to deploy a Wave app on Heroku.
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:14]:
```suggestion
This blog post describes how you can deploy a Wave app on Heroku.
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+<!--truncate-->
+
+Following are the step by step instructions to deploy your [Wave][wave] app to [Heroku][heroku-home] platform. All the code for this blog post is also available in a [GitHub repository][wave-heroku-repo].
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:18]:
```suggestion
Here are step by step instructions to deploy your [Wave][wave] app to [Heroku][heroku-home] platform. All the code for this blog post is also available in a [GitHub repository][wave-heroku-repo].
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+$ python -m venv .venv
+$ . .venv/bin/activate
+(.venv)$ pip install --upgrade pip
+(.venv)$ pip install -r requirements.txt
+```
+
+#### Run the Wave App
+
+Let's run our app on our local machine.
+If you are using Wave version older than `0.20.0`, you need to start the Wave server (`waved`) before running the
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:69]:
```suggestion
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+2022/02/04 19:13:52 #
+2022/02/04 19:13:52 # ┌────────────────┐ H2O Wave
+2022/02/04 19:13:52 # │  ┐┌┐┐┌─┐┌ ┌┌─┐ │ 0.20.0 20220131055911
+2022/02/04 19:13:52 # │  └┘└┘└─└└─┘└── │ © 2021 H2O.ai, Inc.
+2022/02/04 19:13:52 # └────────────────┘
+2022/02/04 19:13:52 #
+```
+
+The app will be available at http://localhost:10101.
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:83]:
```suggestion
The app will be available at http://localhost:10101.


Note that if you are using a Wave version older than `0.20.0`, you'll need to first start the Wave server (`waved`) before running the Python app. Please refer to the [Wave Docs][wave-older-versions].
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+#### Get Dockerfile
+
+For this step we need to add 3 new files to our repo. All three files are available at [srini-x/wave-heroku][wave-heroku-repo].
+Place all three files at the root of the repo/project.
+
+- [Dockerfile][wave-app-dockerfile] - This is the file we need to build our Docker image.
+- [docker-entrypoint.sh][wave-app-docker-entrypoint] - This is the bash script that is called to start out Wave app inside the Docker container.
+- [.dockerignore][wave-app-dockerignore] - This file is similar to `.gitignore` file which contains the list of all unnecessary files.
+
+If you are following along from the top and cloned the example repo, the files are already there.
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:96]:
```suggestion
If you are following along from the top and have cloned the example repo, the files will already be there.
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+2022/02/05 03:34:30 #
+2022/02/05 03:34:30 # {"address":":10101","base-url":"/","t":"listen","web-dir":"/home/appuser/venv/www"}
+INFO:     Started server process [1]
+INFO:     Waiting for application startup.
+2022/02/05 03:34:31 # {"host":"http://127.0.0.1:8000","route":"/","t":"app_add"}
+INFO:     Application startup complete.
+INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
+```
+
+app will be available at http://localhost:10101
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:178]:
```suggestion
Your app can be accessed at http://localhost:10101
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+
+### 3. Deploy the Docker image to Heroku
+
+Ok. Hard part is over. Let's login to Heroku and deploy our app.
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:182]:
```suggestion
OK. The hard part is over. Let's login to Heroku and deploy our app.
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+
+#### Log into Heroku using CLI
+
+Once we have Heroku CLI installed, we can login to Heroku using the following command.
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:186]:
```suggestion
Once we have the Heroku CLI installed, we can login to Heroku using the following command.
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+
+#### Log into Container Registry
+
+```shell
+$ heroku container:login
+```
+
+#### Create a Heroku app
+
+Be in the root of the repo and run.
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:206]:
```suggestion
Navigate to the root of the repo and run.
```

File: website/blog/2022-02-08-wave-on-heroku.md [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+Git URL:        https://git.heroku.com/my-wave-app-1234.git
+Owner:          <your-email@mail.com>
+Region:         us
+Repo Size:      0 B
+Slug Size:      0 B
+Stack:          container
+Web URL:        https://my-wave-app-1234.herokuapp.com/
+```
+
+## What is Next?
```
Comment (lo5) [inline website/blog/2022-02-08-wave-on-heroku.md:280]:
```suggestion
## What's next?
```

**Review by lo5 — CHANGES_REQUESTED** [review]
Thanks a ton for putting this together @srini-x!

Only minor changes - please accept if you're OK with them.

**Comment by mturoci** [comment]
Closing as this has already been published [here](https://medium.com/@srini.x/deploy-your-wave-app-in-5-minutes-bd440d00917f?source=rss-2dbd7cf54686------2)

**Pull request closed without merging**